lib/bundler/definition.rb

```diff
@@ -480,14 +482,21 @@ def specs_changed?(source, &block)
         end
       end
 
-      !locked || unlocking || dependencies_for_source_changed?(locked) || source.specs != locked.specs
+      !locked || unlocking || dependencies_for_source_changed?(source) || specs_for_source_changed?(source)
     end
 
     def dependencies_for_source_changed?(source)
       deps_for_source = @dependencies.select {|s| s.source == source }
       locked_deps_for_source = @locked_deps.select {|s| s.source == source }
 
-      deps_for_source != locked_deps_for_source
+      Set.new(deps_for_source) != Set.new(locked_deps_for_source)
+    end
+
+    def specs_for_source_changed?(source)
+      locked_index = Index.new
+      locked_index.use(@locked_specs.select {|s| source.can_lock?(s) })
+
+      source.specs != locked_index
```

Comment:
I think there's still a bug here because `source.specs` for path gems includes development gems, while locked_specs does not.  I think we need to do something similar to how converge_locked_specs discards dev gems on line 635 below:

 [deps2 = other.dependencies.select {|d| d.type != :development }](https://github.com/bundler/bundler/pull/4618/files#diff-ec267c71117bf68f38c928a3691ba98cR635)  cc @fryguy